### ISSUE TITLE
Embedding layer doesn't support calculate data gradient

### ISSUE THREAD

**Issue by saicoco** [dmlc/mxnet#5258]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
ubuntu15.10
Compiler:
gcc-4.9x
Package used (Python/R/Scala/Julia):
python
MXNet version:
0.9.4
Or if installed from source:
yes
MXNet commit hash (`git rev-parse HEAD`):
be38c5b84030a63d0ab51f19737f99a75a7feb23
If you are using python package, please provide

Python version and distribution:
python2.7
If you are using R package, please provide

R `sessionInfo()`:

## Error Message:
Please paste the full error message, including stack trace.
```
src/operator/tensor/./indexing_op.h:227: Check failed: req[embedding::kData] == kNullOp (3 vs. 0) 
Embedding layer doesn't support calculate data gradient

```
## Minimum reproducible example
if you are using your own code, please provide a short script that reproduces the error.
```python
    cnn_shapes = {
        'image_data': (batch_size, 3, 224, 224)
    }
    cnn_sym = vgg16_fc7('image_data')
    cnn_exec = cnn_sym.simple_bind(ctx=ctx, is_train=False, **cnn_shapes)
    lstm = caption_module(num_lstm_layer=num_lstm_layer, seq_len=train_data.sent_length,
                          vocab_size=train_data.vocab_size, num_hidden=num_hidden, num_embed=num_embed, batch_size=batch_size)
    lstm_shapes = {
        'image_feature': (batch_size, 4096),
        'word_data': (batch_size, train_data.sent_length),
        'softmax_label': (batch_size, train_data.sent_length)
    }

    lstm_exec = lstm.simple_bind(
        ctx=ctx, is_train=True, grad_req='add', **lstm_shapes)

    # init params
    pretrain = mx.nd.load(config.vgg_pretrain)
    init_cnn(cnn_exec, pretrain)

    # init optimazer
    optimazer = mx.optimizer.create('sgd')
    optimazer.lr = learning_rate
    updater = mx.optimizer.get_updater(optimazer)

    # init metric
    perplexity = mx.metric.Perplexity(ignore_label=-1)
    perplexity.reset()

    # callback
    callbacks = collections.namedtuple('callbacks', 'nbatch eval_metric epoch')
    params = callbacks(nbatch=len(train_data.idx)//batch_size, eval_metric=perplexity, epoch=epoches)
    speedometer = mx.callback.Speedometer(batch_size=batch_size, frequent=20)
    for epoch in range(epoches):
        for i, batch in enumerate(train_data):

            # cnn forward, get image_feature
            cnn_exec.arg_dict['image_data'] = batch.data[0]
            cnn_exec.forward()
            image_feature = cnn_exec.outputs[0]

            # lstm forward
            lstm_exec.arg_dict['image_feature'] = image_feature
            lstm_exec.arg_dict['word_data'] = batch.data[1]
            lstm_exec.arg_dict['softmax_label'] = batch.label

            # update lstm grad
            for key, arr in lstm_exec.grad_dict.items():
                arr[:] = 0.
            
            lstm_exec.forward(is_train=True)
            params.eval_metric.update(labels=batch.label,
                              preds=lstm_exec.outputs)
            if 'word_data' in lstm_exec.grad_dict.keys():
                lstm_exec.grad_dict.pop('word_data')
                lstm_exec.arg_dict.pop('word_data')
            lstm_exec.backward()
            speedometer(params)

            for j, name in enumerate(lstm.list_arguments()):
                if name not in lstm_shapes.keys():
                    updater(j, lstm_exec.grad_dict[name], lstm_exec.arg_dict[name])
```
following code is caption_module:  

```python
def caption_module(num_lstm_layer, seq_len, vocab_size, num_hidden, num_embed, batch_size, dropout=0.):

    seq = mx.sym.Variable('word_data')
    label = mx.sym.Variable('softmax_label')
    image_feature = mx.sym.Variable('image_feature')
    image_embed = mx.sym.FullyConnected(data=image_feature, num_hidden=num_embed, name='img_embed')
    word_embed = mx.sym.Embedding(data=seq, input_dim=vocab_size, output_dim=num_embed, name='seq_embed')

    # Concat image_embed with word_embed in axis=1 (batch_size, length+1, num_hidden)
    image_embed_reshape = mx.sym.expand_dims(image_embed, axis=1, name='image_embed_expand_dims')
    embedd_feature = mx.sym.Concat(image_embed_reshape, word_embed, num_args=2, dim=1, name='embed_concat')

    stack = mx.rnn.SequentialRNNCell()
    for i in range(num_lstm_layer):
        stack.add(mx.rnn.LSTMCell(num_hidden=num_hidden, prefix='lstm_l%d_' % i))
    outputs, states = stack.unroll(length=seq_len+1, inputs=embedd_feature, merge_outputs=True)

    outputs_crop = mx.sym.crop(outputs, begin=(0, 1, 0), end=(batch_size, seq_len+1, num_embed))
    pred = mx.sym.Reshape(outputs_crop, shape=(-1, num_hidden))
    pred = mx.sym.FullyConnected(pred, num_hidden=vocab_size, name='pred_fc')
    label = mx.sym.Reshape(label, shape=(-1,))
    softmax_output = mx.sym.SoftmaxOutput(data=pred, label=label, name='softmax')
    return softmax_output
```
I just use pretrain-vgg to extract fc-layer ouputs, and then input them into caption module name `image_feature`, I create custom-dataIter, process them by vgg and then copy them into `image_feature`.
However, it raises above error all the time.
## Steps to reproduce
or if you are running standard examples, please provide the commands you have run that lead to the error.

1.
2.
3.

## What have you tried to solve it?

1.Try to pop item in `grad_dict` about `word_data`
2.Use Module API, and bind symbol.
3.

**Comment by piiswrong** [comment]
Data is discrete. Embedding is non-differentiable wrt data

**Comment by saicoco** [comment]
Thanks, by the way, what should I do to avoid the error? Write embedding layer outside of caption module? 发自网易邮箱大师 在2017年03月06日 04:19，Eric Junyuan Xie 写道： Data is discrete. Embedding is non-differentiable wrt data — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by sxjscience** [comment]
@saicoco Change the line `word_embed = mx.sym.Embedding(data=seq, input_dim=vocab_size, output_dim=num_embed, name='seq_embed')` to
`word_embed = mx.sym.Embedding(data=mx.sym.BlockGrad(seq), input_dim=vocab_size, output_dim=num_embed, name='seq_embed')`

**Comment by saicoco** [comment]
@sxjscience It works~~, I just know `BlockGrad` which can get the outputs of symbol, too naive...`BlockGrad=Block The Grad of Sym`, Thanks  for your second help!
